File: engine/license.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time
from datetime import datetime, timedelta, timezone


def _b64(obj: dict) -> str:
    raw = json.dumps(obj, separators=(",", ":")).encode("utf-8")
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def _unb64(payload: str) -> dict:
    padding = "=" * (-len(payload) % 4)
    raw = base64.urlsafe_b64decode(payload + padding)
    return json.loads(raw)


def _sign(payload: str, secret: str) -> str:
    computed = hmac.new(secret.encode("utf-8"), payload.encode("ascii"), hashlib.sha256).hexdigest()
    return computed[:32]


def make_key(customer: str, months: int, secret: str, plan: str = "monthly") -> str:
    now = datetime.now(timezone.utc)
    issued = now.isoformat()
    expires = (now + timedelta(days=30 * months)).isoformat()
    payload = _b64({"customer": customer, "plan": plan, "issued": issued, "expires": expires})
    return "ATLAS.{}.{}".format(payload, _sign(payload, secret))


def verify_key(key: str, secret: str, now: int | None = None) -> dict:
    now = int(time.time()) if now is None else now
    if not key or not key.startswith("ATLAS."):
        return {"valid": False, "reason": "malformed"}
    parts = key.split(".")
    if len(parts) != 3:
        return {"valid": False, "reason": "malformed"}
    payload, signature = parts[1], parts[2]
    expected = _sign(payload, secret)
    if not hmac.compare_digest(signature, expected):
        return {"valid": False, "reason": "bad signature"}
    try:
        data = _unb64(payload)
        expires_ts = int(datetime.fromisoformat(data["expires"]).timestamp())
    except (KeyError, ValueError, json.JSONDecodeError):
        return {"valid": False, "reason": "corrupt payload"}
    remaining_days = (expires_ts - now) / 86400.0
    if remaining_days <= 0:
        data["remaining_days"] = 0.0
        return {"valid": False, "reason": "expired", "data": data}
    data["remaining_days"] = round(remaining_days, 2)
    return {"valid": True, "data": data, "reason": "ok"}

File: dashboard/lib/license.js
const crypto = require("crypto")

const PREFIX = "ATLAS."

function b64(obj) {
  return Buffer.from(JSON.stringify(obj)).toString("base64url")
}

function un64(payload) {
  return JSON.parse(Buffer.from(payload, "base64url").toString("utf8"))
}

function sign(payload, secret) {
  return crypto.createHmac("sha256", secret).update(payload).digest("hex").slice(0, 32)
}

function generateKey(customer, months, secret, plan = "monthly") {
  const issued = new Date().toISOString()
  const expires = new Date(Date.now() + months * 30 * 24 * 3600 * 1000).toISOString()
  const payload = b64({ customer, plan, issued, expires })
  return `${PREFIX}${payload}.${sign(payload, secret)}`
}

function verifyKey(key, secret) {
  if (typeof key !== "string" || !key.startsWith(PREFIX)) {
    return { valid: false, reason: "malformed" }
  }
  const parts = key.split(".")
  if (parts.length !== 3) return { valid: false, reason: "malformed" }
  const [, payload, signature] = parts
  const expected = sign(payload, secret)
  const a = Buffer.from(signature)
  const b = Buffer.from(expected)
  if (a.length !== b.length || !crypto.timingSafeEqual(a, b)) {
    return { valid: false, reason: "bad signature" }
  }
  let data
  try {
    data = un64(payload)
  } catch {
    return { valid: false, reason: "corrupt payload" }
  }
  const expiresTs = Date.parse(data.expires)
  if (Number.isNaN(expiresTs)) return { valid: false, reason: "corrupt payload", data }
  if (Date.now() > expiresTs) {
    data.remaining_days = 0
    return { valid: false, reason: "expired", data }
  }
  data.remaining_days = Math.round(((expiresTs - Date.now()) / 86400000) * 100) / 100
  return { valid: true, reason: "ok", data }
}

module.exports = { generateKey, verifyKey }